File: multi_head_attention.c
#include "multi_head_attention.h"
#include <stdlib.h>
#include <math.h>
#include <string.h>

// Helper function declarations
static void scaled_dot_product_attention(
    const float* query,  // shape: [batch_size, num_heads, query_seq_len, key_dim]
    const float* key,    // shape: [batch_size, num_heads, key_seq_len, key_dim]
    const float* value,  // shape: [batch_size, num_heads, value_seq_len, value_dim]
    const float* mask,   // shape: [batch_size, query_seq_len, key_seq_len]
    int batch_size,
    int num_heads,
    int query_seq_len,
    int key_seq_len,
    int key_dim,
    int value_dim,
    float* output);      // shape: [batch_size, num_heads, query_seq_len, value_dim]

static void softmax(float* x, int length);

MultiHeadAttention* mha_create(int num_heads, int key_dim, int value_dim) {
    MultiHeadAttention* mha = (MultiHeadAttention*)malloc(sizeof(MultiHeadAttention));
    if (!mha) return NULL;
    
    mha->num_heads = num_heads;
    mha->key_dim = key_dim;
    mha->value_dim = value_dim > 0 ? value_dim : key_dim;
    
    mha->query_weight = NULL;
    mha->key_weight = NULL;
    mha->value_weight = NULL;
    mha->output_weight = NULL;
    
    return mha;
}

void mha_free(MultiHeadAttention* mha) {
    if (mha) {
        free(mha->query_weight);
        free(mha->key_weight);
        free(mha->value_weight);
        free(mha->output_weight);
        free(mha);
    }
}

void mha_init_weights(MultiHeadAttention* mha, int query_dim, int value_dim) {
    if (mha->query_weight) return;  // Weights already initialized
    
    mha->query_dim = query_dim;
    mha->value_dim_in = value_dim;
    
    // Calculate sizes
    int qw_size = query_dim * mha->num_heads * mha->key_dim;
    int kw_size = value_dim * mha->num_heads * mha->key_dim;
    int vw_size = value_dim * mha->num_heads * mha->value_dim;
    int ow_size = (mha->num_heads * mha->value_dim) * query_dim;
    
    // Allocate memory
    mha->query_weight = (float*)malloc(qw_size * sizeof(float));
    mha->key_weight = (float*)malloc(kw_size * sizeof(float));
    mha->value_weight = (float*)malloc(vw_size * sizeof(float));
    mha->output_weight = (float*)malloc(ow_size * sizeof(float));
    
    // Initialize with random values (simplified Xavier initialization)
    float scale = sqrtf(2.0f / (query_dim + mha->key_dim));
    
    // TODO: Initialize weights with proper random values
    // For now, just set to small random values for testing
    for (int i = 0; i < qw_size; i++) mha->query_weight[i] = 0.1f * ((float)rand() / RAND_MAX - 0.5f) * scale;
    for (int i = 0; i < kw_size; i++) mha->key_weight[i] = 0.1f * ((float)rand() / RAND_MAX - 0.5f) * scale;
    for (int i = 0; i < vw_size; i++) mha->value_weight[i] = 0.1f * ((float)rand() / RAND_MAX - 0.5f) * scale;
    for (int i = 0; i < ow_size; i++) mha->output_weight[i] = 0.1f * ((float)rand() / RAND_MAX - 0.5f) * scale;
}

void mha_forward(MultiHeadAttention* mha,
                const float* query,
                const float* value,
                const float* key,
                const float* mask,
                int batch_size,
                int query_seq_len,
                int value_seq_len,
                float* output) {
    // Initialize key if not provided
    if (!key) key = value;
    
    // Initialize weights if needed
    mha_init_weights(mha, query_seq_len, value_seq_len);
    
    // Temporary buffers for intermediate calculations
    // TODO: Implement the forward pass
    // This will involve:
    // 1. Linear transformations for Q, K, V
    // 2. Scaled dot-product attention
    // 3. Output transformation
    
    // For now, just copy input to output for testing
    memcpy(output, query, batch_size * query_seq_len * mha->query_dim * sizeof(float));
}

static void scaled_dot_product_attention(
    const float* query,
    const float* key,
    const float* value,
    const float* mask,
    int batch_size,
    int num_heads,
    int query_seq_len,
    int key_seq_len,
    int key_dim,
    int value_dim,
    float* output) {
    // TODO: Implement scaled dot-product attention
}

static void softmax(float* x, int length) {
    // Find max value for numerical stability
    float max_val = x[0];
    for (int i = 1; i < length; i++) {
        if (x[i] > max_val) max_val = x[i];
    }
    
    // Compute exp and sum
    float sum = 0.0f;
    for (int i = 0; i < length; i++) {
        x[i] = expf(x[i] - max_val);
        sum += x[i];
    }
    
    // Normalize
    for (int i = 0; i < length; i++) {
        x[i] /= sum;
    }
} 

File: multi_head_attention.py
import numpy as np

class MultiHeadAttention:
    def __init__(self, num_heads, key_dim, value_dim=None):
        """Initialize the Multi-Head Attention layer.
        
        Args:
            num_heads: Number of attention heads
            key_dim: Size of each attention head for query and key
            value_dim: Size of each attention head for value (if None, = key_dim)
        """
        self.num_heads = num_heads
        self.key_dim = key_dim
        self.value_dim = value_dim if value_dim else key_dim
        
        # Initialize weights (we'll use random for now, but these should be loaded from trained model)
        self.query_weight = None
        self.key_weight = None
        self.value_weight = None
        self.output_weight = None
        
    def _init_weights(self, query_dim, value_dim):
        """Initialize or verify weights dimensions."""
        if self.query_weight is None:
            # Initialize weights with Xavier/Glorot initialization
            scale = np.sqrt(2.0 / (query_dim + self.key_dim))
            self.query_weight = np.random.normal(0, scale, 
                (query_dim, self.num_heads, self.key_dim))
            self.key_weight = np.random.normal(0, scale, 
                (value_dim, self.num_heads, self.key_dim))
            self.value_weight = np.random.normal(0, scale, 
                (value_dim, self.num_heads, self.value_dim))
            self.output_weight = np.random.normal(0, scale, 
                (self.num_heads * self.value_dim, query_dim))

    def _scaled_dot_product_attention(self, query, key, value, mask=None):
        """Calculate scaled dot-product attention.
        
        Args:
            query: shape (batch_size, num_heads, query_seq_len, key_dim)
            key: shape (batch_size, num_heads, key_seq_len, key_dim)
            value: shape (batch_size, num_heads, value_seq_len, value_dim)
            mask: Optional mask shape (batch_size, query_seq_len, key_seq_len)
        
        Returns:
            attention_output: shape (batch_size, num_heads, query_seq_len, value_dim)
            attention_weights: shape (batch_size, num_heads, query_seq_len, key_seq_len)
        """
        # Calculate attention scores
        matmul_qk = np.matmul(query, np.transpose(key, (0, 1, 3, 2)))
        
        # Scale matmul_qk
        dk = np.float32(self.key_dim)
        scaled_attention_logits = matmul_qk / np.sqrt(dk)
        
        # Apply mask if provided
        if mask is not None:
            # Add extra dimensions for broadcasting
            mask = np.expand_dims(mask, axis=1)  # For num_heads dimension
            scaled_attention_logits += (1 - mask) * -1e9
        
        # Softmax is applied to the last axis
        attention_weights = self._softmax(scaled_attention_logits)
        
        # Calculate output
        output = np.matmul(attention_weights, value)
        
        return output, attention_weights
    
    def _softmax(self, x):
        """Apply softmax to the last dimension."""
        exp_x = np.exp(x - np.max(x, axis=-1, keepdims=True))
        return exp_x / np.sum(exp_x, axis=-1, keepdims=True)
    
    def __call__(self, query, value, key=None, mask=None):
        """Forward pass for the Multi-Head Attention layer.
        
        Args:
            query: shape (batch_size, query_seq_len, query_dim)
            value: shape (batch_size, value_seq_len, value_dim)
            key: Optional, shape (batch_size, key_seq_len, key_dim)
            mask: Optional mask shape (batch_size, query_seq_len, key_seq_len)
            
        Returns:
            output: shape (batch_size, query_seq_len, query_dim)
        """
        if key is None:
            key = value
            
        batch_size = query.shape[0]
        query_seq_len = query.shape[1]
        value_seq_len = value.shape[1]
        
        # Initialize weights if needed
        self._init_weights(query.shape[-1], value.shape[-1])
        
        # Linear transformations and reshape
        # Shape: (batch_size, seq_len, num_heads, dim)
        q = np.einsum('bsd,dhk->bhsk', query, self.query_weight)
        k = np.einsum('bsd,dhk->bhsk', key, self.key_weight)
        v = np.einsum('bsd,dhv->bhsv', value, self.value_weight)
        
        # Calculate attention
        scaled_attention, attention_weights = self._scaled_dot_product_attention(
            q, k, v, mask)
        
        # Reshape and apply output transformation
        # Concatenate heads and apply output transformation
        concat_attention = scaled_attention.transpose(0, 2, 1, 3).reshape(
            batch_size, query_seq_len, self.num_heads * self.value_dim)
        
        output = np.matmul(concat_attention, self.output_weight)
        
        return output


def test_multi_head_attention():
    # Test parameters
    batch_size = 2
    query_seq_len = 4
    key_seq_len = 6
    d_model = 8
    num_heads = 2
    key_dim = 4
    
    # Create test inputs
    query = np.random.normal(0, 1, (batch_size, query_seq_len, d_model))
    value = np.random.normal(0, 1, (batch_size, key_seq_len, d_model))
    
    # Create mask (optional)
    mask = np.ones((batch_size, query_seq_len, key_seq_len))
    
    # Initialize layer
    mha = MultiHeadAttention(num_heads=num_heads, key_dim=key_dim)
    
    # Test forward pass
    output = mha(query, value, mask=mask)
    
    print("Input shapes:")
    print(f"Query shape: {query.shape}")
    print(f"Value shape: {value.shape}")
    print(f"Output shape: {output.shape}")
    
    # Expected output shape: (batch_size, query_seq_len, d_model)
    assert output.shape == (batch_size, query_seq_len, d_model)
    
    return output

# Run the test
if __name__ == "__main__":
    output = test_multi_head_attention()
    print("\nTest passed successfully!") 